### ISSUE TITLE
Multiple output messages (e.g. refresh-all) fails

### ISSUE THREAD

**Issue by djfjeff** [TD22057/insteon-mqtt#86]
Hi,

I want to migrate insteon-mqtt from my Raspberry Pi to a VM. I installed everything with the same procedure that I have, connected the modem to the VM and started insteon-mqtt.

To this point, everything seems to work fine. But if I try to do a refresh-all, I get these error on the terminal : 

```
No read handler found for message type 0x69: All link get first: ack: True
No read handler found for message type 0x57: InpAllLinkRec: 21.xx.xx grp: 1 in_use: True type: CTRL last: False data: 0x01 0x00 0x00
No read handler found for message type 0x69: All link get first: ack: True
```

The weird issue is that after the command finish, insteon-mqtt has erased my modem data file in data/

If I connect the modem back on my Raspberry Pi, refresh-all works fine there.

Any idea?

**Comment by djfjeff** [comment]
I even installed the same version of the OS (Fedora 27) instead of the newer one and still the same issue. I tried installing the same pip package version from the old venv and still the same issue.

I am starting to be short of idea to try to make this work.

**Comment by DCandE** [comment]
Shot in the dark but it might be a file permissions issue. double check how you copied the files and what user you used to copy the files to the new system.

**Comment by djfjeff** [comment]
@DCandE thanks for the pointer but the user is the same user, same uid/gid between both system.

Just to make sure, insteon-mqtt "work" in a sense that I see Insteon message coming in the log and MQTT. So the connection to the modem work.

I just cannot do a refresh-all without having these errors.

**Comment by DCandE** [comment]
Have you tried with the force flag?

**Comment by djfjeff** [comment]
No, I have not. I don't want to risk something on the database because I want to be able to plug back the old system at the end of the day if I cannot make it work on the new.

Also, if I plug back in the old system, the refresh-all command work perfectly.

**Comment by DCandE** [comment]
That read handler is supposed to be accepted by insteon-mqtt/insteon_mqtt/message/OutAllLinkGetFirst.py

**Comment by DCandE** [comment]
are you doing it via the command line `insteon-mqtt config.yaml refresh-all`

**Comment by djfjeff** [comment]
@DCandE yes, I start and do the refresh-all on the command line.

**Comment by djfjeff** [comment]
@DCandE Same result using MQTT to initiate the refresh-all

**Comment by djfjeff** [comment]
Also note, there is some handler that work : 

```
2018-10-19 13:48:04 INFO Protocol: Read 0x62: Std: 41.xx.xx, Type.DIRECT, 19 01 ack: True
2018-10-19 13:48:04 DEBUG Protocol: Passing msg to write handler
2018-10-19 13:48:04 DEBUG DeviceRefresh: 41.xx.xx ACK response
```

**Comment by djfjeff** [comment]
@TD22057 I worked on this over the weekend and was not able to solve this issue. I thought migrating this would be a simple installation/copy the configs!

Any idea, I am running out of thing to try on my end!

**Comment by TD22057** [comment]
It looks to me like there is another process talking to the modem somehow.  Since MQTT is used as the communication, you could have two instances of Insteon-mqtt that both are subscribed to insteon/command so when you send the refresh command, they both start trying to talk and get confused some how.  You also didn't really describe how the VM is working.  Not sure if there are issues passing the serial port through to the VM or not.

You can also do simple manual tests.  Look at notes/test1.py around line 55.  Edit that to have a address of one of your devices.  Then run python -i test1.py and run read() to insure there is nothing waiting.  Then call port.send( msg1 ) and then call read() a few times to see the results.  You should see messages that are similar to what's in test1.py.  If not, there is some issue w/ the serial connection to the PLM (more likely) or in the insteon network (no idea how that would occur).  Look through test1.py and try a few other commands and make sure the comm to the modem is working fine.

If that looks ok, then edit Protocol.py line 220 and uncomment the debug message there.  You can also add another debug line for just the data variable as well.  Make sure you don't have any other insteon-mqtt running somewhere, delete the data storage directory and start it up.  Then try sending the same command you did in test1 (device on/off/etc) and see if you get the same results.

**Comment by djfjeff** [comment]
@TD22057 Thanks for the reply.

I can confirm the VM is not the issue, I tried replicating on a bare metal server and the same issue arise.

I can also confirm that there is no other insteon-mqtt running at the same time and no other MQTT server running too.

Test test1.py works fine (but I need to use port.write instead of port.send). The lights go on and I see the message correctly...

It seems that your code register some handler (we can see 0x50 working) but not some other (0x69 and 0x57).

I confirm that I use the same Python 3.6.6 on both my old server and my new one. Here is the pip list :

```
Package      Version
------------ -------
insteon-mqtt 0.6.3  
Jinja2       2.10   
MarkupSafe   1.0    
paho-mqtt    1.4.0  
pip          18.1   
pyserial     3.4    
PyYAML       3.13   
setuptools   40.4.3
```

**Comment by djfjeff** [comment]
Here is the debug log with the line 220 from Protocol.py uncommented and doing the refresh-all :

```
2018-10-23 08:47:15 DEBUG Protocol: Searching message (len 8): 02 69 06 02 57 e2 01 21... 
2018-10-23 08:47:15 INFO Protocol: Read 0x69: All link get first: ack: True
2018-10-23 08:47:15 DEBUG Protocol: Passing msg to write handler
2018-10-23 08:47:15 WARNING Protocol: No read handler found for message type 0x69: All link get first: ack: True
2018-10-23 08:47:15 DEBUG Mqtt: MQTT publish insteon/command/modem/session/168971936 {"type": "MESSAGE", "data": "No read handler found for message type 0x69: All link get first: ack: True"} qos=0 ret=False
2018-10-23 08:47:15 DEBUG Protocol: Searching message (len 5): 02 57 e2 01 21... 
2018-10-23 08:47:15 DEBUG Mqtt: MQTT writing
2018-10-23 08:47:15 DEBUG Protocol: Searching message (len 18): 02 57 e2 01 21 e9 86 01 00 00 15 02 69 06 02 57 e2 01... 
2018-10-23 08:47:15 INFO Protocol: Read 0x57: InpAllLinkRec: 21.e9.86 grp: 1 in_use: True type: CTRL last: False data: 0x01 0x00 0x00
2018-10-23 08:47:15 DEBUG Protocol: Passing msg to write handler
2018-10-23 08:47:15 WARNING Protocol: No read handler found for message type 0x57: InpAllLinkRec: 21.e9.86 grp: 1 in_use: True type: CTRL last: False data: 0x01 0x00 0x00
2018-10-23 08:47:15 DEBUG Mqtt: MQTT publish insteon/command/modem/session/168971936 {"type": "MESSAGE", "data": "No read handler found for message type 0x57: InpAllLinkRec: 21.e9.86 grp: 1 in_use: True type: CTRL last: False data: 0x01 0x00 0x00"} qos=0 ret=False
```

**Comment by TD22057** [comment]
Did you also run the exact same call in insteon-mqtt as test1.py?.  If the light on/off commands in test1.py work properly, then run the same command in insteon-mqtt and check the logs.

For the refresh problem, that isn't enough log output.  The code does set handlers for messages but it does that only in the correct context.  Meaning the handler for the 0x69 message is activated when that message should show up (after a refresh message is sent) - not at other times.  So the problems isn't in the log section you're posting - it's before that.  If the handler is not active, there is some reason why it was turned off (timed out, something else interrupted it) - that should be in the logs.  Post the full log from when the refresh command request arrives until the end.

ps: I do all my dev and testing on Ubuntu 16.04 so a real 64 bit OS isn't the issue.

**Comment by djfjeff** [comment]
@TD22057 The lights on/off work fine also in insteon-mqtt (sending on or off on the insteon-mqtt command line).

To be honest, there is not a lot of thing before this in the log. However, I just saw something. In my old server, when I do a refresh, I get this at the beginning :

```
2018-08-22 14:28:33 INFO Mqtt: MQTT message insteon/command/modem b'{"cmd": "refresh_all", "force": false, "session": "2515417624"}'
2018-08-22 14:28:33 UI Mqtt: Commanding Modem device 49.f7.73 (modem) cmd=refresh_all
2018-08-22 14:28:33 DEBUG Mqtt: MQTT publish insteon/command/modem/session/2515417624 {"type": "MESSAGE", "data": "Commanding Modem device 49.f7.73 (modem) cmd=refresh_all"} qos=0 ret=False
2018-08-22 14:28:33 INFO Modem: Modem sending get first db record command
2018-08-22 14:28:33 INFO Protocol: Write to modem: All link get first:
2018-08-22 14:28:33 DEBUG Protocol: Write to modem: 02 69
2018-08-22 14:28:33 INFO Base: Device 25.c2.9b (gazebo_light) cmd: status refresh
```

You can see the write 02 69 to the modem. This line is not present on the new server :

```
2018-10-23 10:22:05 INFO Mqtt: MQTT message insteon/command/modem b'{"cmd": "refresh_all", "force": false, "session": "1310778687"}'
2018-10-23 10:22:05 UI Mqtt: Commanding Modem device 49.f7.73 (modem) cmd=refresh_all
2018-10-23 10:22:05 DEBUG Mqtt: MQTT publish insteon/command/modem/session/1310778687 {"type": "MESSAGE", "data": "Commanding Modem device 49.f7.73 (modem) cmd=refresh_all"} qos=0 ret=False
2018-10-23 10:22:05 INFO Modem: Modem sending get first db record command
2018-10-23 10:22:05 INFO Protocol: Write to modem: All link get first:
2018-10-23 10:22:05 INFO Base: Device 25.c2.9b (gazebo_light) cmd: status refresh
```

This is probably why the 0x69 handler is not found later. Any idea why this happen on the new server?

**Comment by djfjeff** [comment]
Ok, latest update. On my Raspberry Pi, I have created a new virtualenv and built the latest version. Guess what, same issue with the handler not found when I do a refresh-all.

**Comment by djfjeff** [comment]
Now I get more issue than only refresh-all. I have the issue when I press a Keypad Linc button to launch a scene in Home Assistant. The scene basically just turn on 2 lights one after the other. Only one turn on or off :

```
2018-10-23 16:52:27 DEBUG Serial: Waiting to write 1540327947.587842 < 1540327947.588033
2018-10-23 16:52:27 DEBUG Serial: Waiting to write 1540327947.587903 < 1540327947.588033
2018-10-23 16:52:27 DEBUG Serial: Waiting to write 1540327947.587985 < 1540327947.588033
2018-10-23 16:52:27 DEBUG Serial: Wrote 8 bytes to serial /dev/insteon
2018-10-23 16:52:27 DEBUG Serial: Wrote 8 bytes to serial /dev/insteon
2018-10-23 16:52:27 INFO Protocol: Read 0x62: Std: 2b.5b.e2, Type.DIRECT, 13 00 ack: False
2018-10-23 16:52:27 DEBUG Protocol: Passing msg to write handler
2018-10-23 16:52:27 DEBUG StandardCmd: 24.89.27 handler unknown msg
2018-10-23 16:52:27 WARNING Protocol: No read handler found for message type 0x62: Std: 2b.5b.e2, Type.DIRECT, 13 00 ack: False
2018-10-23 16:52:27 DEBUG Protocol: No 0x02 starting byte found - clearing
2018-10-23 16:52:27 INFO Protocol: Read 0x50: Std: 42.17.87 Type.ALL_LINK_BROADCAST grp: 04 cmd: 06 00
2018-10-23 16:52:27 DEBUG Protocol: Setting next write time: 1540327947.955759
2018-10-23 16:52:27 DEBUG MsgHistory: Received 0 hops, total 0 for 3 entries
2018-10-23 16:52:27 DEBUG Protocol: Passing msg to write handler
```

**Comment by djfjeff** [comment]
Ok, me again. I think I found something. Using the commit id 1151ef8, everything works perfectly (refresh-all, keypad, etc).

This commit seems to be before all the timing stuff (wait for write, hops count, etc). Could it be that my new server (Intel NUC Core i7) is much faster than my old Raspberry Pi and show timing error?

**Comment by TD22057** [comment]
In theory the processor won't matter at all (I normally run on a full quad core server) but obviously it seems like there is a bug somewhere in that code.  The point of the timing code was to delay sending things so that Insteon behaves more reliably (if you send too fast, bad things happen).  I haven't been running lately so I might not have noticed the problem (I've been busy rewiring my house).  I'll try and take a look this week and see if I can reproduce what's happening.

**Comment by TD22057** [comment]
I'm able to reproduce this.  The refresh-all command is basically broken.  Frankly I'm not sure how it worked at all now that I'm looking at it.  I've pushed a fix to the dev branch which includes and update to allow missing devices to time out and still continue w/ the refresh (the old version when it did work would stop if a device didn't exist).

**Comment by djfjeff** [comment]
@TD22057 thank you! I will try to test this as soon as I can. I just had 3-4 days of unstable home automation here so I think I will let the dust settle a little bit ;)

I know about the refresh-all, it was not the most stable before but running it 2-3 times in a row would work. I now have a work around for this using the simple refresh in a bash script :

```
#!/bin/bash
for i in `cat devices.yaml | grep "-" | awk '{print $2}' | sed 's/://g'`
do
  insteon-mqtt config.yaml refresh $i
done
```

What is impacting me more is the latest timing issue with Keypad and multiple light turning on/off.

**Comment by TD22057** [comment]
FYI - I think these are all related to the same problem.  I thought more about it and while the prev refresh-all wasn't as good as the new one, I think it still should have worked (like it did before).  So I think whatever broke when the delay/hop counting code was added caused refresh all to fail and probably any call that injects multiple commands into the queue.  Once I figure that out, I think that should solve the other problems as well.

**Comment by TD22057** [comment]
There is a bug whenever multiple messages are sent quickly.  The changes for the timing code push a message to the serial link but it doesn't actually write it out until it's safe to do so.  The Protocol still thought it was written though so it tried to write the next message on the next loop which caused the weird message you're seeing.  I have a fix that seems to work but I'm not sure it's a good design or not.  I need to think about where the code really belongs.  Once I get that figured out, I'll submit a fix and push out a 0.64 release.

**Comment by TD22057** [comment]
I believe the last push to the dev branch fixes this.  Feel free to re-open if that's not true.  I'm going to release it as 0.64 so that others can test as well.
